### PR TITLE
Add completion-at-point implementation ocaml-community/utop#261

### PR DESCRIPTION
**Pull request by j-shilling**
After poking around for ways to use my normal `completion-at-point` workflow in the utop buffer, I came accross issue #261, which is a feature request to provide a completion-at-point-fuction.

This is an implementation I came up with that works for me and I wanted to share it here in case anyone else found it useful or had any feedback.

### CODE REVIEW

**Comment by rgrinberg** [comment]
@bbatsov could you review?

File: src/top/utop.el [hunk 1]
```diff
@@ -595,21 +615,43 @@ it is started."
...
                (with-current-buffer utop-complete-buffer
                  (with-output-to-temp-buffer "*Completions*"
                    (display-completion-list (nreverse utop-completion))))
-             (minibuffer-hide-completions))))
+             (minibuffer-hide-completions)))))
        (setq utop-completion nil)))))
 
+(defun utop-completion-at-point ()
+  "Complete thing at point."
+  (when utop-capf-enable
```
Comment (bbatsov) [inline src/top/utop.el:639]: I think this guard here is probably redundant, as you guard for this several places and it's unlikely that people will toggle it off after they loaded the mode.

File: src/top/utop.el [hunk 2]
```diff
@@ -595,21 +615,43 @@ it is started."
...
+    (setq utop-capf-completion nil)
```
Comment (bbatsov) [inline src/top/utop.el:640]: Seems this name is misleading as that's actually the list of completion candidates, right?

**Comment by bbatsov** [comment]
Overall the code looks good to me. In general I think this should be enabled by default - probably unconditionally, given it's the more generic method of completion in Emacs. I see very little point to have to enable manually a `capf` function. Given how basic the company-backend is, probably at some point it can go away, as company has a capf backend anyways.

**Comment by bbatsov** [comment]
While you're at this - I'd also suggest incorporating those doc improvements proposed a while ago https://github.com/ocaml-community/utop/pull/326

**Comment by j-shilling** [comment]
> Overall the code looks good to me. In general I think this should be enabled by default - probably unconditionally, given it's the more generic method of completion in Emacs. I see very little point to have to enable manually a `capf` function. Given how basic the company-backend is, probably at some point it can go away, as company has a capf backend anyways.

As a temporary fix, I added a dispatch function that can be bound to TAB that either calls `completion-at-point` or starts the company workflow. I haven't used company in a long time, but is there any more to this than just removing the current backend and adding `company-capf` backend instead in `utop-mode`?

If it's that simple, I can do that here and then merge in #326 with changes that explain which backend to use. It might also be worth removing the `complete` command and renaming `company-complete` along with removing the `complete-word` response. I'm very new to OCAML, so I might not be the right guy to do that.

**Comment by bbatsov** [comment]
> I haven't used company in a long time, but is there any more to this than just removing the current backend and adding company-capf backend instead in utop-mode?

You don't really need to do anything in utop-mode for `company-capf` to work, provided it registers some completions-at-point function. The purpose of the special company backends is to provide some enhanced functionality, more metadata, etc. Obviously that wasn't the case with the existing backend, that's why I think it becomes kind of redundant. 

> It might also be worth removing the complete command and renaming company-complete along with removing the complete-word response.

Yeah, I'm not sure why it was made company-specific in utop itself. I guess the author of the functionality was a company user and this was before capf became popular (and company alternatives like corfu).

**Comment by bbatsov** [comment]
@rgrinberg @emillon Seems we forgot about this PR, which is good to merge IMO. (especially given that company-mode as not as popular as it used to be)

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**